mp_cnn/dataset.py

```diff
@@ -104,12 +110,36 @@ def initialize(self, word_index, embedding):
             sent_a_tokens.append(sa_tokens)
             sent_b_tokens.append(sb_tokens)
 
+            unique_tokens = set(sa_tokens) | set(sb_tokens)
+            for t in unique_tokens:
+                word_to_doc_cnt[t] += 1
+
         self.sentences = []
+        stoplist = set(stopwords.words('english'))
+        num_docs = len(word_to_doc_cnt)
         for i in range(len(sent_a)):
             sent_pair = {}
             sent_pair['a'] = self._get_sentence_embeddings(sent_a_tokens[i], word_index, embedding)
             sent_pair['b'] = self._get_sentence_embeddings(sent_b_tokens[i], word_index, embedding)
+
+            tokens_a_set, tokens_b_set = set(sent_a_tokens[i]), set(sent_b_tokens[i])
+            intersect = tokens_a_set & tokens_b_set
+            overlap = len(intersect) / (len(tokens_a_set) + len(tokens_b_set))
+            idf_intersect = sum(np.math.log(num_docs / word_to_doc_cnt[w]) for w in intersect)
+            idf_weighted_overlap = idf_intersect / (len(tokens_a_set) + len(tokens_b_set))
+
+            tokens_a_set_no_stop = set(w for w in sent_a_tokens[i] if w not in stoplist)
+            tokens_b_set_no_stop = set(w for w in sent_b_tokens[i] if w not in stoplist)
+            intersect_no_stop = tokens_a_set_no_stop & tokens_b_set_no_stop
+            overlap_no_stop = len(intersect_no_stop) / (len(tokens_a_set_no_stop) + len(tokens_b_set_no_stop))
+            idf_intersect_no_stop = sum(np.math.log(num_docs / word_to_doc_cnt[w]) for w in intersect_no_stop)
+            idf_weighted_overlap_no_stop = idf_intersect_no_stop / (len(tokens_a_set_no_stop) + len(tokens_b_set_no_stop))
+            ext_feats = torch.Tensor([overlap, idf_weighted_overlap, overlap_no_stop, idf_weighted_overlap_no_stop])
+            ext_feats = ext_feats.cuda() if self.cuda else ext_feats
```

Comment:
A small suggestion here - 
```python
if torch.cuda.is_available() and args.cuda:
    torch.cuda.set_device(args.gpu)
```
This would be a better option since you don't have to transfer the`Variable` to cuda

Reply:
Actually check this out. In `main.py` I have this:

```py
args.cuda = not args.no_cuda and torch.cuda.is_available()
```

`args.cuda` is what's passed into the Dataset object and available as `self.cuda`. So the check was already done I think!

Reply:
Ok I see I misunderstood you. Thanks for the tip.